File: Ransom Note - Leetcode 383/Ransom Note - Leetcode 383.java
import java.util.HashMap;

class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        HashMap<Character, Integer> hashmap = new HashMap<>(); 
        
        for (char ch : magazine.toCharArray()) {
            hashmap.put(ch, hashmap.getOrDefault(ch, 0) + 1);
        }

        for (char ch : ransomNote.toCharArray()) {
            if (hashmap.getOrDefault(ch, 0) > 0) {
                hashmap.put(ch, hashmap.get(ch) - 1);
            } else {
                return false;
            }
        }
        
        return true;
    }
}

// Time Complexity: O(m + n)  -> m = len(ransomNote), n = len(magazine)
// Space Complexity: O(n)     -> we're using a hashmap

File: Ransom Note - Leetcode 383/Ransom Note - Leetcode 383.js
var canConstruct = function(ransomNote, magazine) {
    let hashmap = {}; 

    for (let ch of magazine) {
        hashmap[ch] = (hashmap[ch] || 0) + 1;
    }

    for (let ch of ransomNote) {
        if (hashmap[ch] > 0) {
            hashmap[ch]--;
        } else {
            return false;
        }
    }

    return true;
};

// Time Complexity: O(m + n)  -> m = length of ransomNote, n = length of magazine
// Space Complexity: O(n)     -> we're using a hashmap
